### PR TITLE
Closure type annotations

### PR DESCRIPTION
**Pull request by juj**
This enables Closure type casts with `/** @type{castedToType} */(someJavaScriptExpression)`. https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#type-casting

### CODE REVIEW

**Comment by juj** [comment]
This PR adds a localmod to terser to retain parentheses from acorn. Posted a mention at https://github.com/terser/terser/issues/564 to make upstream terser aware of this use case.

**Comment by kripken** [comment]
How does preserving closure comments relate to preserving parens?

What is the closure issue with parens?

I thought terser was used from npm (in which case how can we have a localmod)? Or is it still bundled?

**Comment by juj** [comment]
They relate via the example type annotation construct shown above, that uses both comments and parentheses.

Closure does not have an issue with parentheses, but requires that extra parentheses are used to mark a type cast.

Terser is used as bundled, not from npm.

**Comment by kripken** [comment]
Oh, the parens outside of the comment are part of closure syntax? Interesting...

File: src/shell.js [hunk 1]
```diff
@@ -240,9 +240,9 @@ if (ENVIRONMENT_IS_SHELL) {
 
   if (typeof print !== 'undefined') {
     // Prefer to use print/printErr where they exist, as they usually work better.
-    if (typeof console === 'undefined') console = {};
-    console.log = print;
-    console.warn = console.error = typeof printErr !== 'undefined' ? printErr : print;
+    if (typeof console === 'undefined') console = /** @type{!Console} */({});
+    console.log = /** @type{!function(this:Console, ...*): undefined} */ (print);
+    console.warn = console.error = /** @type{!function(this:Console, ...*): undefined} */ (typeof printErr !== 'undefined' ? printErr : print);
   }
```
Comment (kripken) [inline src/shell.js]: how do annotations like these help in practice?

Comment (kripken) [reply src/shell.js]: Oh, this avoids warnings I guess?

**Comment by juj** [comment]
Yeah, Closure definitely gets rid of excessive parentheses (it'd be quite horrible for it to adopt an annotation syntax that it did not optimize out itself when minifying)

File: src/shell.js (hunk 1 above)
Comment (juj) [reply src/shell.js]: They are to help the type inference, in most cases it is just to mute warnings - on other occassions they can help prevent Closure from making a bad optimization, or to help it do a good one. (examples of both in a later PR)

Comment (juj) [reply src/shell.js]: in this case it is just to mute a warning coming from the assignment.

**Pull request merged**